pf/proto/block.go

```diff
@@ -0,0 +1,134 @@
+// Copyright 2016-2024, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package proto
+
+import (
+	"github.com/hashicorp/terraform-plugin-go/tfprotov6"
+
+	shim "github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfshim"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+)
+
+// Each conceptual block has three components:
+//
+// - blockSchema can be directly returned from a [shim.SchemaMap], as it implements
+// [shim.Schema].
+//
+// - blockResource is the pseudoResource that represents a partial access of the fields in
+// a map. It is only created as the result of [blockSchema.Elem()].
+//
+// - blockMap is the attribute map returned from calling [blockResource.Schema()].
+
+var (
+	_ = shim.Schema(blockSchema{})
+	_ = shim.Resource(blockResource{})
+	_ = shim.SchemaMap(blockMap{})
+)
+
+type blockResource struct {
+	pseudoResource
+	block *tfprotov6.SchemaBlock
+}
+
+func (b blockResource) Schema() shim.SchemaMap     { return blockMap{b.block} }
+func (b blockResource) DeprecationMessage() string { return deprecated(b.block.Deprecated) }
+
+type blockMap struct{ block *tfprotov6.SchemaBlock }
+
+func (m blockMap) Len() int { return len(m.block.Attributes) + len(m.block.BlockTypes) }
+
+func (m blockMap) Get(key string) shim.Schema { return getSchemaMap(m, key) }
+
+func (m blockMap) GetOk(key string) (shim.Schema, bool) {
+	for _, a := range m.block.Attributes {
+		if a.Name == key {
+			return attribute{*a}, true
+		}
+	}
+
+	for _, b := range m.block.BlockTypes {
+		if b.TypeName == key {
+			return blockSchema{*b}, true
+		}
+	}
+
+	return nil, false
+}
+
+func (m blockMap) Range(each func(key string, value shim.Schema) bool) {
+	for _, a := range m.block.Attributes {
+		if !each(a.Name, attribute{*a}) {
+			return
+		}
+	}
+
+	for _, b := range m.block.BlockTypes {
+		if !each(b.TypeName, blockSchema{*b}) {
+			return
+		}
+	}
+}
+
+func (m blockMap) Set(key string, value shim.Schema) {
```

Comment:
Curious if we can panic here and be OK with an immutable model. Does it get used?

Reply:
I'd rather not panic. If we don't need these methods, let's remove them from `shim.SchemaMap`. I'll explore that in a seperate PR.